### PR TITLE
[GPU] Implement ExperimentalDetectronGenerateProposalsSingleImage-6

### PR DESCRIPTION
**Pull request by tgubanova-lohika**
ExperimentalDetectronGenerateProposalsSingleImage-6 implementation with unit and single-layer tests.

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/ops/experimental_detectron_generate_proposals_single_image.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                       tensor_from_dims(output_shape)};
+    cldnn::memory::ptr shared_memory{p.GetEngine().allocate_memory(mutable_layout)};
+
+    const auto mutable_id_w = layer_type_name + "_md_write";
+    const cldnn::mutable_data mutable_prim_w{mutable_id_w, shared_memory, op_friendly_name};
+    p.primitiveIDs[mutable_id_w] = mutable_id_w;
+    p.AddPrimitive(mutable_prim_w);
+    inputs.push_back(mutable_id_w);
+
+    const cldnn::experimental_detectron_generate_proposal_single_image prim{layer_name,
```
Comment (mhnap-lohika) [inline src/plugins/intel_gpu/src/plugin/ops/experimental_detectron_generate_proposals_single_image.cpp:50]: Seems that there was a typo in cldnn op's name, no? Cause factory name is ExperimentalDetectronGenerateProposal**s**SingleImage, but cldnn op's name without 's' in "proposal" word. Or this is intended?

**Comment by yeonbok** [comment]
Original PR: https://github.com/openvinotoolkit/openvino/pull/10023

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            arguments.push_back({ArgumentDescriptor::Types::INTERNAL_BUFFER, 1}); // nms_out_indices
+            arguments.push_back({ArgumentDescriptor::Types::INTERNAL_BUFFER, 2}); // nms_num_outputs
+            break;
+        }
+        case 3: { // convert proposals to rois and roi scores
+            arguments.push_back({ArgumentDescriptor::Types::INTERNAL_BUFFER, 0}); // proposals
+            arguments.push_back({ArgumentDescriptor::Types::INTERNAL_BUFFER, 1}); // nms_out_indices
+            arguments.push_back({ArgumentDescriptor::Types::INTERNAL_BUFFER, 2}); // nms_num_outputs
+            arguments.push_back({ArgumentDescriptor::Types::OUTPUT, 0});          // rois
+            arguments.push_back({ArgumentDescriptor::Types::INPUT, kRoiScoresOutputIdx}); // roi scores
```
Comment (opoluektov-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:91]: Could you please explain why roi scores argument is marked as INPUT and not OUTPUT?

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            }
+        }
+    }
+}
+
+// 1. Sort boxes by scores
+KERNEL(edgpsi_ref_stage_1)(__global INPUT0_TYPE* proposals) {
+    __global Box* boxes = (__global Box*)proposals;
+
+    FUNC_CALL(quickSortIterative)(boxes, 0, NUM_PROPOSALS-1);
```
Comment (opoluektov-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:177]: if `pre_nms_count` is expected to be significantly smaller than total number of proposals, using partial sort here could improve performance quite a bit...

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    smoke_ExperimentalDetectronGenerateProposalsSingleImageLayerTest,
+    ExperimentalDetectronGenerateProposalsSingleImageLayerTest,
+    ::testing::Combine(
+        ::testing::ValuesIn(input_shape),
+        ::testing::ValuesIn(min_size),
+        ::testing::ValuesIn(nms_threshold),
+        ::testing::ValuesIn(post_nms_count),
+        ::testing::ValuesIn(pre_nms_count),
+        ::testing::ValuesIn(input_tensors),
+        ::testing::Values(ov::element::Type_t::f32),
```
Comment (wilson-seok) [inline src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp:113]: Could you please add test case of f16 precision? In the specification, the op can support any supported floating point.. https://docs.openvino.ai/latest/openvino_docs_ops_detection_ExperimentalDetectronGenerateProposalsSingleImage_6.html

File: src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        }
+    }
+};
+
+TEST_P(experimental_detectron_generate_proposals_single_image_test, basic) {
+    ASSERT_NO_FATAL_FAILURE(test());
+}
+
+INSTANTIATE_TEST_SUITE_P(
+        experimental_detectron_generate_proposals_single_image_gpu_test,
```
Comment (wilson-seok) [inline src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp:162]: Could you please add test case for fp16?

**Comment by wilson-seok** [comment]
fp16 model compilation failed during kernel compiling. You can reproduce the issue from conformance test.
$./conformanceTests --gtest_filter=*OpImplCheck*ExperimentalDetectronGenerateProposalsSingleImage* --device=GPU
Or try ReadIR test with f16 IR files.

image

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/edgpsi_kernel_ref.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                cldnn_jit.AddConstants({MakeJitConstant("POST_NMS_COUNT", new_params.post_nms_count)});
+                break;
+            }
+            default:
+                throw std::invalid_argument("EDGPSI has 4 kernels. valid index is 0 ~ 3.");
+        }
+
+        const auto jit = CreateJit(kernelName, cldnn_jit, entry_point);
+        auto& kernel = kd.kernels[i];
+        KernelBase::CheckDispatchData(kernelName, dispatchData, params.engineInfo.maxWorkGroupSize);
```
[the remark this answers is not in the document]
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/edgpsi_kernel_ref.cpp]: not exactrly, kernel arguments are filled in other way. I don't think it make sense to change base class impl now

File: src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    auto inputs = p.GetInputPrimitiveIDs(op);
+
+    const auto& attrs = op->get_attrs();
+
+    const auto op_friendly_name = op->get_friendly_name();
+
+    const auto layer_type_name = layer_type_name_ID(op);
+    const auto layer_name_refine_anchors = layer_type_name + "_refine_anchors";
```
[the remark this answers is not in the document]
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp (hunk 2 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:91]: no, here should be INPUT, as we return indices as input mutable data

File: src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp (hunk 5 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp:245]: done

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp (hunk 4 above)
Comment (tgubanova-lohika) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp]: done

File: src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp]: done

File: src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    const auto layer_name = layer_type_name + ".0";
+
+
+
```
[the remark this answers is not in the document]
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/plugin/ops/edgpsi.cpp]: done

File: src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp (hunk 5 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/test_cases/experimental_detectron_generate_proposals_single_image_gpu_test.cpp:245]: Thanks!

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp (hunk 4 above)
Comment (wilson-seok) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/experimental_detectron_generate_proposals_single_image.cpp]: Thanks!

**Comment by wilson-seok** [comment]
> fp16 model compilation failed during kernel compiling. You can reproduce the issue from conformance test. $./conformanceTests --gtest_filter=_OpImplCheck_ExperimentalDetectronGenerateProposalsSingleImage* --device=GPU Or try ReadIR test with f16 IR files.
> 
> image

f16 model loading is now passed after the commit 9557747.

File: src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+protected:
+    kernel_arguments_data get_arguments(typed_primitive_inst<experimental_detectron_generate_proposals_single_image>& instance, int32_t) const override {
+        kernel_arguments_data args;
+        const auto num_inputs = instance.inputs_memory_count();
+        for (size_t i = 0; i < num_inputs; ++i) {
+            args.inputs.push_back(instance.input_memory_ptr(i));
+        }
+
+        args.outputs.push_back(instance.output_memory_ptr());
+        args.inputs.push_back(instance.output_roi_scores_memory());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp:33]: Note:  All implementation is assuming that the second output memory is created always and used. 
For performance, I think the second input should be created and used only when it is used. 
If we dont care of performance matter since there is only a ref kernel, it is okay for now. 
However in the future if we need to optimize the memory usage and performance, we need to make it to use second output only when it is needed....

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    FUNC_CALL(swap_box)(&arr[i + 1], &arr[h]);
+    return (i + 1);
+}
+
+inline void FUNC(bubbleSortIterative)(__global Box* arr, int l, int h) {
+    for (int i = 0; i < h - l; i++) {
+        bool swapped = false;
+        for (int j = l; j < h - i; j++) {
+            if ((arr[j].score > arr[j + 1].score)) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: The arr[j].score == arr[j+1].score case is not covered. In the refernece code (src/core/reference/include/ngraph/runtime/reference/experi....hpp), the sort is being done with std::sort, 
which will be different for the platform library. 
In high probability this undefined equal case is not a problem for CPU plugin because both CPU plugin and ngraph plugin are using same sort (possibly). However in GPU we define our own sort function as here, and the inputs order is random, such "equal score case"s result might not be aligned with the std::sort in some platforms. As a result, the functional tests in multiple platforms might fail randomly. 
So in conclusion we need to make all of the plugins to be aligned with same comparator, which might need to change all of referece impl, cpu plugin kernel impl, and gpu kernel here... 
Please refer to DetectionOutput's experience. https://github.com/openvinotoolkit/openvino/pull/6530

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: how do you think, should it be done in scope or this PR?

File: src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp (hunk 9 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp:33]: agree with you. can we make it in the future as an improvement?

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl (hunk 10 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: I think it should be applied before this PR is merged.
Also it is better to be applied as a separate PR from this PR for better review process, since it will need to touch other components (ngraph, cpu plugin). 
Otherwise it will make our CI to fail randomly after this PR is merged.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: It is a common issue for all GPU operations which implements own sort. There just a few for now, NMS and DetectionOutput, and the issues are not existing for those ops because they implemented handling of same score cases. Now you are adding this new ops without handling of same score cases, we need to add that hadnling before merging this PR. Otherwise we will experience random failure of our CI in the future.

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: understood. So for now  I think the best case would be to override generation as it's done in DetectionOutput. Would it be acceptable for this PR?

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: actually now inputs are already received from params, but I can change it to generation

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+namespace kernel_selector {
+
+ParamsKey ExperimentalDetectronGenerateProposalsSingleImageRef::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::F16);
+    k.EnableInputDataType(Datatype::F32);
+    k.EnableOutputDataType(Datatype::F16);
+    k.EnableOutputDataType(Datatype::F32);
+    k.EnableInputLayout(DataLayout::bfyx);
+    k.EnableOutputLayout(DataLayout::bfyx);
```
Comment (wilson-seok) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:19]: <html>
<body>
<p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; color: rgb(201, 209, 217); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Could you please add those blocked format for input/output layout? Those are required for dGPU support.</p>

Prec | Layout
-- | --
int8 | fsv16, fsv32, bsv32_fsv32
fp16 | fsv16, bsv32_fsv16
fp32 | fsv16, bsv32_fsv16


</body>
</html>

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:19]: we should add it to all the open PRs, right?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:19]: Yes, please add them for all PR if it doesn't have.

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:21]: done

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:21]: Sorry for confusion. I checked internally and we need full support of  fsv16, fsv32, bsv32_fsv16, bsv32_fsv32, bsv16_fsv16, bsv16_fsv32. I will update comments in other PRs.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl (hunk 10 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: @yeonbok  as inputs are not random and we don't expect sporadic fails because of this can we handle this conversation as resolved?

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp (hunk 11 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:24]: done

File: src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp (hunk 9 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/graph/impls/ocl/experimental_detectron_generate_proposals_single_image.cpp:35]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl (hunk 10 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:124]: Oh, this test is not using random input as others. Then it is okay to go with. Thanks for checking.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp (hunk 11 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:24]: If we enabled them, we'll need correspodning unittests. Please add them.

**Comment by wilson-seok** [comment]
Please rebase the PR to resolve issue in openvino-mac check.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp (hunk 11 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/edgpsi/experimental_detectron_generate_proposals_single_image_kernel_ref.cpp:19]: as we discussed today - removed this layouts for now

**Review by wilson-seok — APPROVED** [review]
LGTM

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl (hunk 3 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:185]: @tgubanova-lohika Is there any on-going work to resolve this?

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:185]: not for now, i propose to leave in ref impl as it is

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_generate_proposals_single_image_ref.cl:185]: we can rework it when will use common sorting for GPU
